src/main/java/org/apache/sysds/runtime/privacy/finegrained/FineGrainedPrivacyList.java

```diff
@@ -35,11 +35,61 @@
 
 	private ArrayList<Map.Entry<DataRange, PrivacyLevel>> constraintCollection = new ArrayList<>();
 
+	@Override
+	public PrivacyLevel[] getRowPrivacy(int numRows, int numCols) {
+		PrivacyLevel[] privacyLevels = new PrivacyLevel[numRows];
+		for (int i = 0; i < numRows; i++) {
+			long[] beginRange1 = new long[] {i, 0};
+			long[] endRange1 = new long[] {i, numCols};
+			privacyLevels[i] = getStrictestPrivacyLevel(new DataRange(beginRange1, endRange1));
+		}
+		return privacyLevels;
+	}
+
+	@Override
+	public PrivacyLevel[] getColPrivacy(int numRows, int numCols) {
+		PrivacyLevel[] privacyLevels = new PrivacyLevel[numCols];
+		for (int i = 0; i < numCols; i++) {
+			long[] beginRange = new long[] {0, i};
+			long[] endRange = new long[] {numRows, i};
+			privacyLevels[i] = getStrictestPrivacyLevel(new DataRange(beginRange, endRange));
+		}
+		return privacyLevels;
+	}
+
+	private PrivacyLevel getStrictestPrivacyLevel(DataRange searchRange){
+		PrivacyLevel strictestLevel = PrivacyLevel.None;
+		for ( Map.Entry<DataRange, PrivacyLevel> constraint : constraintCollection ) {
```

Comment:
because you have a for loop in the previous and loop through each entry here. it is going to be slow.
I think a future project would be an index structure for the privacy. maybe a tree structure would be good?
one for cols and one for rows?

Reply:
I actually started implementing a tree structure when I added fine-grained privacy, but we decided that it was not relevant for now and that we could do it later when we have the propagation of fine-grained constraints ready.

Reply:
same jira task as mentioned above